### PR TITLE
Back-end services

### PR DESCRIPTION
**Pull request by srinathsetty**
This pull request contains an end-to-end implementation the back-end services for the following REST APIs of the blockchain service layer: /transactions, /blocks/:hash, and /blocks/last.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,11 +35,13 @@
   },
   "dependencies": {
     "@types/bitcore-lib": "^0.15.1",
+    "@types/node-fetch": "^2.1.2",
     "@types/raw-body": "^2.3.0",
     "@types/request-promise-native": "^1.0.15",
```
Comment (thehenrytsai) [inline package.json]: Sanity: Remove `request-promise-native` dependency if not used. Same for line 47.

File: src/RequestHandler.ts [hunk 2]
```diff
@@ -42,33 +43,35 @@ export default class RequestHandler {
    * @param tx Sidetree transaction to write into the underlying blockchain.
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Rename `tx` to `anchorFileHash` for consistency throughout the repo.

File: src/RequestHandler.ts [hunk 3]
```diff
@@ -42,33 +43,35 @@ export default class RequestHandler {
...
    */
   public async handleAnchorRequest (tx: string): Promise<Response> {
-    const baseUrl = 'http://http://104.40.11.171:3001/SidetreeRooterService';
+    const prefix = 'sidetreetest:';
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Consider using a hardcoded value that will be used in production, such as just `Sidetree`, else make this configurable.

File: src/RequestHandler.ts [hunk 4]
```diff
@@ -42,33 +43,35 @@ export default class RequestHandler {
...
-      'c83820b6016814e8dd3b5ca416b118fd0ba8c66857bdd3d788b3ef85307d80121023eb8129ab5cae93cb940f74dd06897' +
-      'df9f20ac6e7ec6ad5ca41e6a35d0415b3fffffffff020000000000000000196a1773696465747265653a68656c6c6f5f7' +
-      '76f726c645f7473b829b90a000000001976a914fc7054837bfb4c3a6fbc2d543567bdb0158b0d5188ac00000000';
-
-    const options = {
-      uri: baseUrl + queryString
+    const uri = baseUrl + queryString;
+    const requestParameters = {
+      method: 'get',
+      headers: { 'Content-Type': 'application/json' }
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: there shouldn't be any content type for a GET request?

File: src/RequestHandler.ts [hunk 5]
```diff
@@ -42,33 +43,35 @@ export default class RequestHandler {
...
     };
 
+    const content = await nodeFetch(uri, requestParameters);
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: What happens if `nodeFetch(...)` fails?

File: src/RequestHandler.ts [hunk 6]
```diff
@@ -42,33 +43,35 @@ export default class RequestHandler {
...
     let response: Response;
 
-    try {
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: I feel it is still best to wrap processing logic with try-catch.

File: src/RequestHandler.ts [hunk 7]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
...
+    }
+
+    return response;
+
+
+
+
+
+
     try {
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: where is there code after return on line 57?

File: src/RequestHandler.ts [hunk 8]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
...
+    } else {
+      response = {
+        status: ResponseStatus.ServerError,
+        body: { 'error': 'Error anchoring Sidetree transaction on Bitcoin' }
+      };
+    }
+
+    return response;
+
+
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: I would have thought that linter we prevent you from having multiple consecutive empty lines.

File: src/RequestHandler.ts [hunk 9]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
...
 
+
+    const prefix = 'sidetreetest:';
+    const baseUrl = 'http://104.40.11.171:3001/SidetreeBlockchainService';
+    const queryString = '/transactions/' + after + '/' + prefix;
+
+    const uri = baseUrl + queryString;
+    const requestParameters = {
+      method: 'get',
+      headers: { 'Content-Type': 'application/json' }
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Sanity: there shouldn't be any content type for a GET request?

File: src/RequestHandler.ts [hunk 10]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
       };
     }
 
+
+    const prefix = 'sidetreetest:';
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Consider using a hardcoded value that will be used in production, such as just `sidetree`, and create a class level property to remove dupliation at line 86, else make this configurable.

File: src/RequestHandler.ts [hunk 11]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
...
+          'moreTransactions': contentBody['moreTransactions'],
+          'blockNumber': contentBody['blockNumber'],
+          'blockHash': contentBody['blockHash'],
+          'hashes': contentBody['hashes']
+        }
+      };
+    } else {
+      response = {
+        status: ResponseStatus.ServerError,
+        body: { 'error': 'Error anchoring Sidetree transaction on Bitcoin' }
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: The error message appears to be incorrect.

File: src/RequestHandler.ts [hunk 12]
```diff
@@ -20,6 +21,46 @@ export default class RequestHandler {
...
+
+    const content = await nodeFetch(uri, requestParameters);
+    let response: Response;
+
+    if (content.status === HttpStatus.OK) {
+      const responseBodyString = (content.body.read() as Buffer).toString();
+      const contentBody = JSON.parse(responseBodyString);
+      response = {
+        status: ResponseStatus.Succeeded,
+        body: {
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: The body does not follow the spec.

File: src/RequestHandler.ts [hunk 13]
```diff
@@ -20,102 +20,146 @@ export default class RequestHandler {
...
-    // TODO: construct the transaction dynamically; the ts version of bitcore-lib has issues
-    const queryString = '/anchor/01000000012dec15faf88ec573e6aef24d534cb252a2c1570fc6525350989a15447ec7' +
-      'c92d010000006b483045022100800f8de83c5f7ebfade178b7a013e71f71cc3c972f19c905a3283ba3307ee1cd0220745' +
-      'c83820b6016814e8dd3b5ca416b118fd0ba8c66857bdd3d788b3ef85307d80121023eb8129ab5cae93cb940f74dd06897' +
-      'df9f20ac6e7ec6ad5ca41e6a35d0415b3fffffffff020000000000000000196a1773696465747265653a68656c6c6f5f7' +
-      '76f726c645f7473b829b90a000000001976a914fc7054837bfb4c3a6fbc2d543567bdb0158b0d5188ac00000000';
-
-    const options = {
-      uri: baseUrl + queryString
+  public async handleAnchorRequest(tx: string): Promise<Response> {
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Linter should have picked this up, space needed in front of `(tx...`

File: src/RequestHandler.ts [hunk 14]
```diff
@@ -20,102 +20,146 @@ export default class RequestHandler {
...
-    try {
-      const result = request.get(options);
+    if (content.status === HttpStatus.OK) {
+      const responseBodyString = (content.body.read() as Buffer).toString();
+      const contentBody = JSON.parse(responseBodyString);
       response = {
         status: ResponseStatus.Succeeded,
-        body: { txId: result, metadata: tx }
+        body: {
+          'blockNumber': contentBody['transactionId']
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Santy: the implementation spec doesn't require anything to be returned.

File: src/RequestHandler.ts [hunk 15]
```diff
@@ -20,102 +20,146 @@ export default class RequestHandler {
       };
     }
 
-    try {
+    const prefix = 'sidetreetest:';
+    const baseUrl = 'http://104.40.11.171:3001/SidetreeBlockchainService';
```
Comment (thehenrytsai) [inline src/RequestHandler.ts]: Also declare `baseUrl` as a class levle property and remove all duplications, optionally make this configurable.

File: tests/RequestHandler.spec.ts [hunk 16]
```diff
@@ -8,15 +8,55 @@ describe('RequestHandler', () => {
     requestHandler = new RequestHandler();
   });
 
+  /*it('should return the correct response body with height for blocks last request', async () => {
```
Comment (thehenrytsai) [inline tests/RequestHandler.spec.ts]: Remove all commented out code.

File: tests/RequestHandler.spec.ts [hunk 17]
```diff
@@ -8,15 +8,55 @@ describe('RequestHandler', () => {
...
+        'blockHash': '00000000000000c0e5f7ef7f5e7ff8bd61b41e04db4035ec7be67a59c50cbb64'
+      }
+    };
+
+    const fetchedResponse = await requestHandler.handleLastBlockRequest();
+    console.info(fetchedResponse);
+    expect(expectedResponse).toEqual(fetchedResponse);
+  });*/
+
+  it('should return the correct response body with height for blocks request', async () => {
```
Comment (thehenrytsai) [inline tests/RequestHandler.spec.ts]: The added tests appear to be integration tests (which will be useful too), but unit tests are what's normally expected, there are more work to be done even if these integration tests are to be accepted. Such as:

1. Test data injection.
1. bitcored endpoint configuration.

At very least create issues to track this.

File: src/RequestHandler.ts (hunk 6 above)
Comment (srinathsetty) [reply src/RequestHandler.ts]: Done.

File: package.json [hunk 18]
```diff
@@ -17,40 +17,40 @@
...
     "koa-router": "^7.2.1",
+    "node-fetch": "^2.3.0",
     "raw-body": "^2.3.3",
-    "request": "^2.88.0",
-    "request-promise-native": "^1.0.5"
+    "request": "^2.88.0"
   },
   "scripts": {
+    "precommit": "npm run validate",
+    "build": "tsc && copyfiles json/*.json dist",
```
Comment (thehenrytsai) [inline package.json]: `copyfiles` is not a dependency.

**Pull request merged**